client/asset/btc/electrum/network.go

```diff
@@ -0,0 +1,646 @@
+// This code is available on the terms of the project LICENSE.md file,
+// also available online at https://blueoakcouncil.org/license/1.0.0.
+
+package electrum
+
+import (
+	"bufio"
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"strconv"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/decred/go-socks/socks"
+)
+
+const pingInterval = 10 * time.Second
+
+// ServerConn represents a connection to an Electrum server e.g. ElectrumX. It
+// is a single use type that must be replaced if the connection is lost. Use
+// ConnectServer to construct a ServerConn and connect to the server.
+type ServerConn struct {
+	conn   net.Conn
+	cancel context.CancelFunc
+	done   chan struct{}
+	proto  string
+
+	reqID uint64
+
+	respHandlersMtx sync.Mutex
+	respHandlers    map[uint64]chan *response // reqID => requestor
+
+	ntfnHandlersMtx sync.RWMutex
+	ntfnHandlers    map[string][]chan []byte // method => subscribers
+}
+
+func (sc *ServerConn) nextID() uint64 {
+	return atomic.AddUint64(&sc.reqID, 1)
+}
+
+const newline = byte('\n')
+
+func (sc *ServerConn) listen(ctx context.Context) {
+	defer sc.cancelRequests() // close the response chans
+	defer sc.deleteSubscriptions()
+
+	reader := bufio.NewReader(io.LimitReader(sc.conn, 1<<18))
+
+	for {
+		if ctx.Err() != nil {
+			return
+		}
+		msg, err := reader.ReadBytes(newline)
+		if err != nil {
+			if ctx.Err() == nil { // unexpected
+				fmt.Println(err)
+			}
+			sc.cancel()
+			return
+		}
+
+		var jsonResp response
+		err = json.Unmarshal(msg, &jsonResp)
+		if err != nil {
+			fmt.Println("Unmarshal error:", err)
+			continue
+		}
+
+		if jsonResp.Method != "" { // notification
+			var ntfnParams ntfnData // the ntfn payload is in the params field of a request object (!)
+			err = json.Unmarshal(msg, &ntfnParams)
+			if err != nil {
+				fmt.Println("Unmarshal error:", err)
+				continue
+			}
+			for _, c := range sc.subChans(jsonResp.Method) {
+				select {
+				case c <- ntfnParams.Params:
+				default:
+				}
```

Comment:
I'm not sure but do you need to hold the `ntfnHandlers` here? You hold it when getting with `subChans` but it looks like they could be closed while sending here.

Reply:
They are closed in the defer of this function, thus synchronized with these sends.  I'm adding docs to both `deleteSubscriptions` and `cancelRequests` to make this clearer.
The rule of thumb with closing channels is to only close them in the goroutine that _sends_ on the channel, which is why I had a comment on the defer.  It can be clearer though.

Reply:
Ok I see, `deleteSubscriptions` is only used at the end of this function in the defer.